src/neml2/solvers/NonlinearSystem.cxx

```diff
@@ -188,4 +188,14 @@ NonlinearSystem::residual_and_Jacobian()
   }
 }
 
+BatchTensor
+NonlinearSystem::residual_norm() const
+{
+  // If the residual is a logical scalar just return its absolute value
+  if (residual_view().base_dim() == 0)
+    return math::abs(residual_view());
+
+  return BatchTensor(torch::linalg::vector_norm(residual_view(), 2, -1, false, c10::nullopt),
```

Comment:
I've been meaning to ask: why was all the ancillary options needed?

Also: we should probably refactor this into `BatchTensorBase` and test it.

Reply:
> I've been meaning to ask: why was all the ancillary options needed?

Because this is c++ not python... This could just be a `math::sqrt(math::bvv(r, r))`.

I agree we should put this into `BatchTensorBase`.